components/raftstore/src/store/worker/split_controller.rs

```diff
@@ -0,0 +1,537 @@
+    pub right: i32,
+}
+
+impl Sample {
+    fn new(key: &[u8]) -> Sample {
+        Sample {
+            key: key.to_owned(),
+            left: 0,
+            contained: 0,
+            right: 0,
+        }
+    }
+}
+
+// It will return prefix sum of iter. `read` is a function to be used to read data from iter.
+fn prefix_sum<F, T>(iter: Iter<T>, read: F) -> Vec<usize>
+    where
+        F: Fn(&T) -> usize,
+{
+    let mut pre_sum = vec![];
+    let mut sum = 0;
+    for item in iter {
+        sum += read(&item);
+        pre_sum.push(sum);
+    }
+    pre_sum
+}
+
+// It will return sample_num numbers by sample from lists.
+// The list in the lists has the length of N1, N2, N3 ... Np ... NP in turn.
+// Their prefix sum is pre_sum and we can get mut list from lists by get_mut.
+// Take a random number d from [1, N]. If d <N1, select a data in the first list with a equal probability without replacement;
+// If N1 <= d <(N1 + N2), then select a data in the second list with equal probability without replacement;
+// and so on, repeat m times, and finally select sample_num pieces of data from lists.
+fn sample<F, T>(
+    sample_num: usize,
+    pre_sum: &[usize],
+    mut lists: Vec<T>,
+    get_mut: F,
+) -> Vec<KeyRange>
+    where
+        F: Fn(&mut T) -> &mut Vec<KeyRange>,
+{
+    let mut rng = rand::thread_rng();
+    let mut key_ranges = vec![];
+    let high_bound = pre_sum.last().unwrap();
+    for _num in 0..sample_num {
+        let d = rng.gen_range(0, *high_bound) as usize;
+        let i = match pre_sum.binary_search(&d) {
+            Ok(i) => i,
+            Err(i) => i,
+        };
+        let list = get_mut(&mut lists[i]);
+        let j = rng.gen_range(0, list.len()) as usize;
+        key_ranges.push(list.remove(j)); // Sampling without replacement
+    }
+    key_ranges
+}
+
+// RegionInfo will maintain key_ranges with sample_num length by reservoir sampling.
+// And it will save qps num and peer.
+#[derive(Debug, Clone)]
+pub struct RegionInfo {
+    pub sample_num: usize,
+    pub qps: usize,
+    pub peer: Peer,
+    pub key_ranges: Vec<KeyRange>,
+}
+
+impl RegionInfo {
+    fn new(sample_num: usize) -> RegionInfo {
+        RegionInfo {
+            sample_num,
+            qps: 0,
+            key_ranges: Vec::with_capacity(sample_num),
+            peer: Peer::default(),
+        }
+    }
+
+    fn get_qps(&self) -> usize {
+        self.qps
+    }
+
+    fn get_key_ranges_mut(&mut self) -> &mut Vec<KeyRange> {
+        &mut self.key_ranges
+    }
+
+    fn add_key_ranges(&mut self, key_ranges: Vec<KeyRange>) {
+        for key_range in key_ranges {
+            if self.key_ranges.len() < self.sample_num {
+                self.key_ranges.push(key_range);
+            } else {
+                let i = rand::thread_rng().gen_range(0, self.qps) as usize;
+                if i < self.sample_num {
+                    self.key_ranges[i] = key_range;
+                }
+            }
+            self.qps += 1;
```

Comment:
Why `qps += 1` ?

Reply:
`self.qps += key_ranges.len()` is better.